### ISSUE TITLE
Null when user cancelled the scan using Back Button

### ISSUE THREAD

**Issue by alvinvin00** [wu9007/qrcode_scanner#109]
Hello there, i'm having some problems that

~~a. If user are pressing Back Button (hardware), Null will be returned instead, the problem is scan method expects Future<String> which means anything that nullable will throw NullError.~~ (EDIT : Solved in 0.3.2)

b. If user are pressing Back Button (software, available on ZKing's platform view), the future will never complete

**Comment by MerdanDev** [comment]
Yes problem b. is very serius. I am having this problem in which I need to return String and when I press back button it will stuck forever. Only solution for android is remove back button because we have hardware back button or software back button, if ios we need to throw error and catch it after that return something.

**Comment by alvinvin00** [comment]
what if user cancels QR Scanning process? Should scan method return null in that case?

**Comment by MerdanDev** [comment]
I think it will throw error. After that we can catch error and we can show that user cancelled scanning.

**Comment by mmkumr** [comment]
Check out this https://pub.dev/packages/barcode_scan_fix.

**Comment by alvinvin00** [comment]
bruh, that package are not null-safe

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by mmkumr** [comment]
I have tried adding the scan function inside try-catch. But still, it fails. So, can you add try-catch in the function?

**Comment by MerdanDev** [comment]
> I have tried adding the scan function inside try-catch. But still, it fails. So, can you add try-catch in the function?

Can you show your code please? Yes it is possible to add try-catch in the function.

**Comment by alvinvin00** [comment]
> I have tried adding the scan function inside try-catch. But still, it fails. So, can you add try-catch in the function?

```
try {
    //Scan QR Code
    code = await _qr.scanFromCamera();
} on TypeError {
    //Catch that pesky null thrown from library and do whatever the heck you want
}
```

This is what i did to deal with them

**Comment by alvinvin00** [comment]
so, as of 0.3.2, problem A has been solved but problem B are still there, the soft back button wasn't removed in the latest release

**Comment by wu9007** [comment]
Fixed / kept on `master` (qrscan 0.4.0).

`scan()` is `Future<String?>`. Back button / Close returns `null`. Denied camera throws `PlatformException` with code `PERMISSION_NOT_GRANTED`.

Closing as completed.

**Comment by wu9007** [comment]
@alvinvin00 splitting your two reports, because 0.3.2 only fixed one of them.

**(a) hardware back → null / TypeError** — fixed in 0.3.2. On Dart 3 the signature is `Future<String?> scan()`, so a cancel is `null`, not a `TypeError`. You no longer need the `on TypeError` catch.

**(b) on-screen `scan_back` never completing the Future** — that was still open after 0.3.2. On 0.4.0 both the overlay back button and `onBackPressed()` call `cancel()` → `RESULT_CANCELED` → Dart `null`. iOS Close does the same.

```dart
final code = await scanner.scan();
if (code == null) {
  // user cancelled — Future completed
}
```

Closing (a) and (b).
